src/sinks/new_relic/config.rs

```diff
@@ -0,0 +1,190 @@
+#[serde(rename_all = "snake_case")]
+#[derivative(Default)]
+pub enum NewRelicRegion {
+    #[derivative(Default)]
+    Us,
+    Eu,
+}
+
+#[derive(Deserialize, Serialize, Debug, Eq, PartialEq, Clone, Copy, Derivative)]
+#[serde(rename_all = "snake_case")]
+#[derivative(Default)]
+pub enum NewRelicApi {
+    #[derivative(Default)]
+    Events,
+    Metrics,
+    Logs
+}
+
+#[derive(Clone, Copy, Debug, Default)]
+pub struct NewRelicDefaultBatchSettings;
+
+impl SinkBatchSettings for NewRelicDefaultBatchSettings {
+    const MAX_EVENTS: Option<usize> = Some(100);
+    const MAX_BYTES: Option<usize> = Some(1_000_000);
+    const TIMEOUT_SECS: NonZeroU64 = unsafe { NonZeroU64::new_unchecked(1) };
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NewRelicApiRetry;
+
+impl RetryLogic for NewRelicApiRetry {
+    type Error = NewRelicSinkError;
+    type Response = NewRelicApiResponse;
+
+    fn is_retriable_error(&self, _error: &Self::Error) -> bool {
+        // Never retry.
+        false
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct NewRelicConfig {
+    pub license_key: String,
+    pub account_id: String,
+    pub region: Option<NewRelicRegion>,
+    pub api: NewRelicApi,
+    #[serde(default = "Compression::gzip_default")]
+    pub compression: Compression,
+    #[serde(
+        skip_serializing_if = "crate::serde::skip_serializing_if_default",
+        default
+    )]
+    pub encoding: EncodingConfigFixed<Encoding>,
+    #[serde(default)]
+    pub batch: BatchConfig<NewRelicDefaultBatchSettings>,
+    #[serde(default)]
+    pub request: TowerRequestConfig
+}
+
+impl_generate_config_from_default!(NewRelicConfig);
+
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "new_relic")]
+impl SinkConfig for NewRelicConfig {
+    async fn build(
+        &self,
+        cx: SinkContext,
+    ) -> crate::Result<(super::VectorSink, super::Healthcheck)> {
+        let encoding = self.encoding.clone();
+
+        let batcher_settings = self.batch
+            .validate()?
+            .limit_max_events(self.batch.max_events.unwrap_or(50))?
+            .into_batcher_settings()?;
+
+        let request_limits = self.request.unwrap_with(&Default::default());
+        let tls_settings = TlsSettings::from_options(&None)?;
+        let client = HttpClient::new(tls_settings, &cx.proxy)?;
+
+        let service = ServiceBuilder::new()
+            .settings(request_limits, NewRelicApiRetry)
+            .service(NewRelicApiService {
+                client
+            });
+
+        let credentials = Arc::from(NewRelicCredentials::from(self));
+
+        let sink = NewRelicSink {
+            service,
+            acker: cx.acker(),
+            encoding,
+            credentials,
+            compression: self.compression,
+            batcher_settings,
+        };
+
+        Ok((
+            super::VectorSink::Stream(Box::new(sink)),
+            future::ok(()).boxed(),
```

Comment:
Can you add a healthcheck? You can look at the existing new relic sink for an example. This is useful to check that the config is correct, and Vector can talk to the API.

Reply:
Regarding health checks, the New Relic APIs don't document a way to check the status using the API itself (they have a status page, but not something you could request with an API), there is no documented GET, HEAD, or any other method besides the POST used to send the actual data. Although using the POST method with an empty object to check the availability is probably possible, is highly discouraged, because it could contribute to reach the API request limits, degrading the service performance.

https://docs.newrelic.com/docs/data-apis/ingest-apis/introduction-event-api/
https://docs.newrelic.com/docs/data-apis/ingest-apis/metric-api/introduction-metric-api/
https://docs.newrelic.com/docs/logs/log-api/introduction-log-api/

Reply:
The healthcheck is only ran a single time when Vector starts up, so I wouldn't worry about API request limits. (And at least for the Events API, you can send 100000 / minute), so if that is your only concern I'd suggest still adding _some_ kind of health check.

The current `new_relic_logs` sends an empty `HEAD` request and checks the status code response, but if an empty `POST` also works, that could be reasonable too.

Reply:
Sounds fair. Only one concern, what if the health check fails? As you said the `new_relic_log` sink uses a HEAD call, but this method is not documented in the official New Relic API descriptions, so, one would expect that this could suddenly change or even disappear. What would happen in such a case? The sink would continue working or Vector will prevent it from running?

Reply:
If the healthcheck fails that specific sink won't run. However, if something breaks the healthcheck it can easily be disabled (eg: https://vector.dev/docs/reference/configuration/sinks/new_relic_logs/#healthcheck.enabled)

Reply:
Implemented a healthcheck in https://github.com/vectordotdev/vector/pull/9764/commits/9c8e7b22f3477ee7f3a31bf2511b79cef2da7ca5

First of all, I want to apologize for a misleading statement I made. I said that New Relic doesn't provide a status API, and that's not true, they actually [does](https://status.newrelic.com/api). So I used this API to check the availability of the corresponding New Relic API endpoints.

Reply:
I think there is a bit of misunderstanding here on the purpose of the healthcheck. This is something that is ran once when Vector starts and is mostly there to verify that the configuration of Vector is correct and it's able to talk to the service. So it should be able to detect issues such as invalid credentials, or invalid network configuration (can't reach the endpoint). Checking the `status` endpoint does verify part of this, but leaves out checking the credentials. Ideally this would check an endpoint that is actually used by the sink (such as sending an empty POST to upload events, if that is a valid request) so that it is testing everything.
We are working on updating the component spec documentation to further clarify the health checks to make this more clear in the future.

Reply:
For more info about health checks, see https://github.com/vectordotdev/vector/blob/master/docs/DEVELOPING.md#sink-healthchecks